### Import the libraries

In [1]:
import pandas as pd
import numpy as np

### Load data

In [2]:
agg_trans = pd.read_csv("aggregated_transaction.csv")
agg_user = pd.read_csv("aggregated_user.csv")
map_trans = pd.read_csv("map_transaction.csv")
map_user = pd.read_csv("map_user.csv")
top_trans = pd.read_csv("top_transaction.csv")
top_user = pd.read_csv("top_user.csv")

agg_trans.head(), agg_user.head()

(   Unnamed: 0                     states  year  quarter  \
 0           0  andaman-&-nicobar-islands  2018        1   
 1           1  andaman-&-nicobar-islands  2018        1   
 2           2  andaman-&-nicobar-islands  2018        1   
 3           3  andaman-&-nicobar-islands  2018        1   
 4           4  andaman-&-nicobar-islands  2018        1   
 
                  trans_type  trans_counts        amount  
 0  Recharge & bill payments          4200  1.845307e+06  
 1     Peer-to-peer payments          1871  1.213866e+07  
 2         Merchant payments           298  4.525072e+05  
 3        Financial Services            33  1.060142e+04  
 4                    Others           256  1.846899e+05  ,
    Unnamed: 0                     states  year  quarter    brand  user_counts  \
 0           0  andaman-&-nicobar-islands  2018        1   Xiaomi         1665   
 1           1  andaman-&-nicobar-islands  2018        1  Samsung         1445   
 2           2  andaman-&-nicobar-isl

### Check shape + column info

In [3]:
datasets = {
    "agg_trans": agg_trans,
    "agg_user": agg_user,
    "map_trans": map_trans,
    "map_user": map_user,
    "top_trans": top_trans,
    "top_user": top_user
}

for name, df in datasets.items():
    print(f"\n--- {name} ---")
    print(df.shape)
    print(df.dtypes)


--- agg_trans ---
(3594, 7)
Unnamed: 0        int64
states           object
year              int64
quarter           int64
trans_type       object
trans_counts      int64
amount          float64
dtype: object

--- agg_user ---
(6732, 7)
Unnamed: 0       int64
states          object
year             int64
quarter          int64
brand           object
user_counts      int64
percentage     float64
dtype: object

--- map_trans ---
(14636, 7)
Unnamed: 0        int64
states           object
year              int64
quarter           int64
district         object
trans_counts      int64
amount          float64
dtype: object

--- map_user ---
(14640, 6)
Unnamed: 0                 int64
states                    object
year                       int64
quarter                    int64
district                  object
registered_user_counts     int64
dtype: object

--- top_trans ---
(5920, 7)
Unnamed: 0        int64
states           object
year              int64
quarter           int64
district

### Check missing values & duplicates

In [4]:
for name, df in datasets.items():
    print(f"\n--- {name} ---")
    print("Null values:\n", df.isnull().sum())
    print("Duplicate rows:", df.duplicated().sum())


--- agg_trans ---
Null values:
 Unnamed: 0      0
states          0
year            0
quarter         0
trans_type      0
trans_counts    0
amount          0
dtype: int64
Duplicate rows: 0

--- agg_user ---
Null values:
 Unnamed: 0     0
states         0
year           0
quarter        0
brand          0
user_counts    0
percentage     0
dtype: int64
Duplicate rows: 0

--- map_trans ---
Null values:
 Unnamed: 0      0
states          0
year            0
quarter         0
district        0
trans_counts    0
amount          0
dtype: int64
Duplicate rows: 0

--- map_user ---
Null values:
 Unnamed: 0                0
states                    0
year                      0
quarter                   0
district                  0
registered_user_counts    0
dtype: int64
Duplicate rows: 0

--- top_trans ---
Null values:
 Unnamed: 0      0
states          0
year            0
quarter         0
district        0
trans_counts    0
amount          0
dtype: int64
Duplicate rows: 0

--- top_user ---

### Drop the ‘Unnamed: 0’ column from all datasets

In [5]:
for name, df in datasets.items():
    if "Unnamed: 0" in df.columns:
        df.drop(columns=["Unnamed: 0"], inplace=True)

# Check again
for name, df in datasets.items():
    print(name, df.columns)

agg_trans Index(['states', 'year', 'quarter', 'trans_type', 'trans_counts', 'amount'], dtype='object')
agg_user Index(['states', 'year', 'quarter', 'brand', 'user_counts', 'percentage'], dtype='object')
map_trans Index(['states', 'year', 'quarter', 'district', 'trans_counts', 'amount'], dtype='object')
map_user Index(['states', 'year', 'quarter', 'district', 'registered_user_counts'], dtype='object')
top_trans Index(['states', 'year', 'quarter', 'district', 'trans_counts', 'amount'], dtype='object')
top_user Index(['states', 'year', 'quarter', 'district', 'registered_user_counts'], dtype='object')


### Standardize text formatting

In [6]:
def clean_text(col):
    return col.str.lower().str.strip().str.replace("&", "and").str.replace("-", " ").str.replace("_", " ")

# Apply to relevant columns
agg_trans["states"] = clean_text(agg_trans["states"])
agg_user["states"] = clean_text(agg_user["states"])
map_trans["states"] = clean_text(map_trans["states"])
map_user["states"] = clean_text(map_user["states"])
top_trans["states"] = clean_text(top_trans["states"])
top_user["states"] = clean_text(top_user["states"])

map_trans["district"] = clean_text(map_trans["district"])
map_user["district"] = clean_text(map_user["district"])
top_trans["district"] = clean_text(top_trans["district"])
top_user["district"] = clean_text(top_user["district"])

# Quick check
agg_trans.head()

states  year  quarter                trans_type  \
0  andaman and nicobar islands  2018        1  Recharge & bill payments   
1  andaman and nicobar islands  2018        1     Peer-to-peer payments   
2  andaman and nicobar islands  2018        1         Merchant payments   
3  andaman and nicobar islands  2018        1        Financial Services   
4  andaman and nicobar islands  2018        1                    Others   

   trans_counts        amount  
0          4200  1.845307e+06  
1          1871  1.213866e+07  
2           298  4.525072e+05  
3            33  1.060142e+04  
4           256  1.846899e+05

### Duplicate & Outlier Verification

#### Check duplicates based on key columns

In [7]:
# Function to check duplicates based on key columns
duplicate_checks = {
    "agg_trans": ["states", "year", "quarter", "trans_type"],
    "agg_user": ["states", "year", "quarter", "brand"],
    "map_trans": ["states", "district", "year", "quarter"],
    "map_user": ["states", "district", "year", "quarter"],
    "top_trans": ["states", "district", "year", "quarter"],
    "top_user": ["states", "district", "year", "quarter"]
}

for name, keys in duplicate_checks.items():
    df = datasets[name]
    dups = df.duplicated(subset=keys).sum()
    print(f"{name}: {dups} duplicates based on {keys}")

agg_trans: 0 duplicates based on ['states', 'year', 'quarter', 'trans_type']
agg_user: 0 duplicates based on ['states', 'year', 'quarter', 'brand']
map_trans: 0 duplicates based on ['states', 'district', 'year', 'quarter']
map_user: 0 duplicates based on ['states', 'district', 'year', 'quarter']
top_trans: 0 duplicates based on ['states', 'district', 'year', 'quarter']
top_user: 0 duplicates based on ['states', 'district', 'year', 'quarter']


#### Outlier detection using Interquartile Range (IQR)
We check numeric fields:

trans_counts

amount

user_counts

registered_user_counts

In [8]:
import numpy as np

numeric_cols = {
    "agg_trans": ["trans_counts", "amount"],
    "agg_user": ["user_counts", "percentage"],
    "map_trans": ["trans_counts", "amount"],
    "map_user": ["registered_user_counts"],
    "top_trans": ["trans_counts", "amount"],
    "top_user": ["registered_user_counts"]
}

for name, cols in numeric_cols.items():
    df = datasets[name]
    print(f"\n--- Outliers in {name} ---")
    for col in cols:
        Q1 = df[col].quantile(0.25)
        Q3 = df[col].quantile(0.75)
        IQR = Q3 - Q1
        lower = Q1 - 1.5 * IQR
        upper = Q3 - 1.5 * IQR
        outliers = df[(df[col] < lower) | (df[col] > upper)]
        print(f"{col}: {len(outliers)} outliers")


--- Outliers in agg_trans ---
trans_counts: 3594 outliers
amount: 3594 outliers

--- Outliers in agg_user ---
user_counts: 6732 outliers
percentage: 6732 outliers

--- Outliers in map_trans ---
trans_counts: 14636 outliers
amount: 14636 outliers

--- Outliers in map_user ---
registered_user_counts: 14640 outliers

--- Outliers in top_trans ---
trans_counts: 5920 outliers
amount: 5920 outliers

--- Outliers in top_user ---
registered_user_counts: 5920 outliers


#### ✅ Understanding the Outlier Results
Why did every value show as an "outlier"?

Because:

PhonePe transaction data varies wildly between states/districts

Some states have lakhs of transactions, some have a few hundred

IQR is not suitable for skewed, non-normal, high-variance financial data

In such datasets, everything becomes an outlier because distribution is not even

So this output means:

✔️ These are not wrong values
✔️ These are natural variations in real digital payment volumes
✔️ We do NOT remove or cap them
🛑 Important

Removing outliers in financial transaction data is incorrect and will completely distort the results.

Examples:

Maharashtra has ~100M+ transactions

Lakshadweep may have <10K
IQR will mark Maharashtra as “outlier” — which is wrong.

🟢 Conclusion

No duplicates

No missing values

Outliers are expected and should NOT be removed

Data is clean and ready for merging

### Create Master Table 1 (State-level)

We will merge:

aggregated_transaction

aggregated_user

on:

states, year, quarter

#### Merge state-level tables

In [10]:
master_state = pd.merge(
    agg_trans,
    agg_user,
    on=["states", "year", "quarter"],
    how="inner"
)

master_state.head()

states  year  quarter                trans_type  \
0  andaman and nicobar islands  2018        1  Recharge & bill payments   
1  andaman and nicobar islands  2018        1  Recharge & bill payments   
2  andaman and nicobar islands  2018        1  Recharge & bill payments   
3  andaman and nicobar islands  2018        1  Recharge & bill payments   
4  andaman and nicobar islands  2018        1  Recharge & bill payments   

   trans_counts        amount    brand  user_counts  percentage  
0          4200  1.845307e+06   Xiaomi         1665    0.247033  
1          4200  1.845307e+06  Samsung         1445    0.214392  
2          4200  1.845307e+06     Vivo          982    0.145697  
3          4200  1.845307e+06     Oppo          501    0.074332  
4          4200  1.845307e+06  OnePlus          332    0.049258

#### Check merged table details

In [11]:
print(master_state.shape)
master_state.sample(5)

(33594, 9)


states  year  quarter  \
33344                               west bengal  2021        1   
17597                               lakshadweep  2021        4   
17879                            madhya pradesh  2018        4   
16333                                    ladakh  2019        4   
6626   dadra and nagar haveli and daman and diu  2018        2   

                     trans_type  trans_counts        amount     brand  \
33344  Recharge & bill payments      26673733  1.133967e+10      Oppo   
17597     Peer-to-peer payments          7154  4.469667e+07  Motorola   
17879  Recharge & bill payments       7273502  1.669346e+09    Realme   
16333                    Others           576  1.253585e+06    Lenovo   
6626          Merchant payments          3544  4.685536e+06    Realme   

       user_counts  percentage  
33344      1829470    0.108840  
17597           46    0.008159  
17879       294612    0.054909  
16333          433    0.010045  
6626          3991    0.055851

#### ✔ Why this merge looks correct?

Because:

Each state-year-quarter combination has many brands

So after merging, each transaction row is duplicated for each brand

This is intentional because PhonePe Pulse aggregated user data is brand-wise

Example:
If a quarter has 5 brands → then each transaction type appears 5 times.

This is normal and expected.

### 🎯 Master Table 1 is ready.

### Create Master Table 2 (District-Level)
This will merge:

map_transaction

map_user

Keys:
states, district, year, quarter
### Merge district-level tables

In [12]:
master_district = pd.merge(
    map_trans,
    map_user,
    on=["states", "district", "year", "quarter"],
    how="inner"
)

master_district.head()

states  year  quarter  \
0  andaman and nicobar islands  2018        1   
1  andaman and nicobar islands  2018        1   
2  andaman and nicobar islands  2018        1   
3  andaman and nicobar islands  2018        2   
4  andaman and nicobar islands  2018        2   

                            district  trans_counts        amount  \
0  north and middle andaman district           442  9.316631e+05   
1             south andaman district          5688  1.256025e+07   
2                  nicobars district           528  1.139849e+06   
3  north and middle andaman district           825  1.317863e+06   
4             south andaman district          9395  2.394824e+07   

   registered_user_counts  
0                     632  
1                    5846  
2                     262  
3                     911  
4                    8143

#### Check district table

In [13]:
print(master_district.shape)
master_district.sample(5)

(14636, 7)


states  year  quarter             district  trans_counts  \
13422   uttar pradesh  2021        2      bijnor district       2371176   
8851         nagaland  2019        3   zunheboto district         20054   
7190   madhya pradesh  2022        1        guna district       5024607   
5618        karnataka  2020        1  kalaburagi district       2328594   
5466        karnataka  2018        4      mysuru district        992071   

             amount  registered_user_counts  
13422  5.226003e+09                  409603  
8851   6.055400e+07                    2058  
7190   8.067959e+09                  343862  
5618   4.631841e+09                  402676  
5466   1.934192e+09                  341636

### Create Master Table 3 (Top Districts)
We will merge:

top_transaction

top_user

Keys:
states, district, year, quarter
### Merge top-level tables

In [14]:
master_top = pd.merge(
    top_trans,
    top_user,
    on=["states", "district", "year", "quarter"],
    how="inner"
)

master_top.head()

states  year  quarter                  district  \
0  andaman and nicobar islands  2018        1             south andaman   
1  andaman and nicobar islands  2018        1                  nicobars   
2  andaman and nicobar islands  2018        1  north and middle andaman   
3  andaman and nicobar islands  2018        2             south andaman   
4  andaman and nicobar islands  2018        2                  nicobars   

   trans_counts        amount  registered_user_counts  
0          5688  1.256025e+07                    5846  
1           528  1.139849e+06                     262  
2           442  9.316631e+05                     632  
3          9395  2.394824e+07                    8143  
4          1120  3.072437e+06                     351

### Check top table

In [15]:
print(master_top.shape)
master_top.sample(5)

(5511, 7)


states  year  quarter   district  trans_counts        amount  \
2567        kerala  2022        2   thrissur       8767715  1.385099e+10   
3659      nagaland  2021        1   tuensang         23643  8.451075e+07   
808   chhattisgarh  2018        2   bilaspur        227499  3.244496e+08   
2959   maharashtra  2022        1     nagpur      43072958  6.306581e+10   
1260           goa  2019        1  south goa        297005  5.344590e+08   

      registered_user_counts  
2567                  808727  
3659                    8866  
808                    94888  
2959                 2286011  
1260                   90350

## Final Master Table (Unified Model for SQL & Power BI)

Now we will prepare a final unified structure for analysis.
1. master_state_year_quarter

(From aggregated tables)

2. master_district_year_quarter

(Your district-level table)

3. master_top_districts

(Your top-level table)

4. dim_state

List of unique states

5. dim_district

List of unique districts

6. fact_transactions

A single fact table for metrics:

transaction_counts

transaction_amount

registered_users      

## ✅ Step A) Python Code for — Final Unified Dataset

Below is the full, clean, production-ready code to generate the final unified analytics dataset using your three merged tables:

master_agg (state-level)

master_district (district-level)

master_top (top-level districts)

This script will:

✔ Standardize column names
✔ Create dimension tables
✔ Create unified fact table
✔ Ensure referential keys (state_id, district_id)
✔ Prepare outputs for SQLite in Step B

In [17]:
# -----------------------------
# Step 8: Final Unified Dataset
# -----------------------------
# Helper cleaners
def clean_state_name(s):
    return s.strip().lower().replace("&", "and").replace("-", " ")

def clean_district_name(s):
    return s.strip().lower()

# Clean text formatting
master_state['states'] = master_state['states'].apply(clean_state_name)

master_district['states'] = master_district['states'].apply(clean_state_name)
master_district['district'] = master_district['district'].apply(clean_district_name)

master_top['states'] = master_top['states'].apply(clean_state_name)
master_top['district'] = master_top['district'].apply(clean_district_name)

# -----------------------------
# Create final unified dataset
# -----------------------------
final_master = {
    "state_level": master_state,
    "district_level": master_district,
    "top_districts": master_top
}

# Display shapes
for name, df in final_master.items():
    print(f"{name}: {df.shape}")

state_level: (33594, 9)
district_level: (14636, 7)
top_districts: (5511, 7)


## Step (B): SQL Database Setup (SQLite)

1. Create SQLite database (phonepe_pulse.db)
2. Create tables and schema
3. Insert the three datasets

You can run all of it inside the notebook — no external tool needed.
## ✅ Step B1 — SQL Schema + Create SQLite Tables

In [18]:
import sqlite3

# Create / connect to SQLite DB
conn = sqlite3.connect("phonepe_pulse.db")
cursor = conn.cursor()

# -----------------------------
# Create Tables
# -----------------------------

cursor.execute("""
CREATE TABLE IF NOT EXISTS state_level (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    states TEXT,
    year INTEGER,
    quarter INTEGER,
    trans_type TEXT,
    trans_counts INTEGER,
    amount REAL,
    brand TEXT,
    user_counts INTEGER,
    percentage REAL
);
""")

cursor.execute("""
CREATE TABLE IF NOT EXISTS district_level (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    states TEXT,
    year INTEGER,
    quarter INTEGER,
    district TEXT,
    trans_counts INTEGER,
    amount REAL,
    registered_user_counts INTEGER
);
""")

cursor.execute("""
CREATE TABLE IF NOT EXISTS top_districts (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    states TEXT,
    year INTEGER,
    quarter INTEGER,
    district TEXT,
    trans_counts INTEGER,
    amount REAL,
    registered_user_counts INTEGER
);
""")

conn.commit()
print("Tables created successfully.")

Tables created successfully.


## ✅ Step B2 — Insert Pandas DataFrames into SQLite

In [19]:
master_state.to_sql('state_level', conn, if_exists='append', index=False)
master_district.to_sql('district_level', conn, if_exists='append', index=False)
master_top.to_sql('top_districts', conn, if_exists='append', index=False)

print("Data inserted into all tables.")

Data inserted into all tables.


## ✔️ After This

### I will have a complete SQLite database:

#### phonepe_pulse.db

with tables:

state_level

district_level

top_districts

Ready for:

SQL queries

Power BI connection

Python analytics

API development

Dashboard creation

###### ✔️ Cleaned datasets
###### ✔️ Final merged master tables
###### ✔️ SQLite database (phonepe_pulse.db)
###### ✔️ Tables fully loaded with data

### I can now move to the analysis layer.

In [24]:
import sqlite3
import pandas as pd

conn = sqlite3.connect("phonepe.db")   # or your actual DB name
print("✅ Connected to phonepe.db")

# List all tables
tables = pd.read_sql("SELECT name FROM sqlite_master WHERE type='table';", conn)
tables

✅ Connected to phonepe.db


Empty DataFrame
Columns: [name]
Index: []

##### ✅ Let's verify tables correctly

In [25]:
pd.read_sql("SELECT name, type FROM sqlite_master;", conn)

Empty DataFrame
Columns: [name, type]
Index: []

### database is empty

No tables exist in phonepe.db yet.

#### Confirm DataFrames exist

In [26]:
print(master_state.shape)
print(master_district.shape)
print(master_top.shape)

(33594, 9)
(14636, 7)
(5511, 7)


##### Now I can proceed properly.

✅ Recreate all SQLite tables + insert all data

In [27]:
import sqlite3

# Connect to DB
conn = sqlite3.connect("phonepe.db")
cursor = conn.cursor()

print("✅ Connected to phonepe.db")

# 1️⃣ Drop old tables if they exist
cursor.executescript("""
DROP TABLE IF EXISTS state_trans_user;
DROP TABLE IF EXISTS district_transactions;
DROP TABLE IF EXISTS top_transactions;
""")

print("🗑️ Old tables cleared")

# 2️⃣ Create fresh tables
cursor.executescript("""
CREATE TABLE state_trans_user (
    states TEXT,
    year INTEGER,
    quarter INTEGER,
    trans_type TEXT,
    trans_counts INTEGER,
    amount REAL,
    brand TEXT,
    user_counts INTEGER,
    percentage REAL
);

CREATE TABLE district_transactions (
    states TEXT,
    year INTEGER,
    quarter INTEGER,
    district TEXT,
    trans_counts INTEGER,
    amount REAL,
    registered_user_counts INTEGER
);

CREATE TABLE top_transactions (
    states TEXT,
    year INTEGER,
    quarter INTEGER,
    district TEXT,
    trans_counts INTEGER,
    amount REAL,
    registered_user_counts INTEGER
);
""")

print("📌 Fresh tables created")

# 3️⃣ Insert DataFrames into SQLite
master_state.to_sql("state_trans_user", conn, if_exists="append", index=False)
master_district.to_sql("district_transactions", conn, if_exists="append", index=False)
master_top.to_sql("top_transactions", conn, if_exists="append", index=False)

print("📥 Data inserted successfully!")

# 4️⃣ Check tables created
tables = cursor.execute("SELECT name FROM sqlite_master WHERE type='table';").fetchall()
print("\n📌 Tables in DB:")
for t in tables:
    print(" -", t[0])

✅ Connected to phonepe.db
🗑️ Old tables cleared
📌 Fresh tables created
📥 Data inserted successfully!

📌 Tables in DB:
 - state_trans_user
 - district_transactions
 - top_transactions


### ⭐ Query 1 — Total Transaction Amount per State

In [28]:
query1 = """
SELECT 
    states,
    SUM(amount) AS total_transaction_amount
FROM state_trans_user
GROUP BY states
ORDER BY total_transaction_amount DESC;
"""

q1_result = pd.read_sql(query1, conn)
q1_result.head()

states  total_transaction_amount
0       telangana              9.775625e+13
1     maharashtra              9.467711e+13
2       karnataka              9.280427e+13
3  andhra pradesh              8.024605e+13
4       rajasthan              5.788288e+13

### ⭐ Query 2 — Total Transaction Count per State

In [29]:
query2 = """
SELECT 
    states,
    SUM(trans_counts) AS total_transaction_count
FROM state_trans_user
GROUP BY states
ORDER BY total_transaction_count DESC;
"""

q2_result = pd.read_sql(query2, conn)
q2_result.head()

states  total_transaction_count
0     maharashtra              63874603694
1       karnataka              62963141087
2       telangana              50960361859
3  andhra pradesh              36497797963
4       rajasthan              31193662500

### ⭐ Query 3 — Top 10 States by Total User Count

In [30]:
query3 = """
SELECT 
    states,
    SUM(user_counts) AS total_users
FROM state_trans_user
GROUP BY states
ORDER BY total_users DESC
LIMIT 10;
"""

q3_result = pd.read_sql(query3, conn)
q3_result

states  total_users
0     maharashtra   2260375055
1   uttar pradesh   1779848165
2       karnataka   1456863900
3  andhra pradesh   1127074175
4       rajasthan   1078227940
5       telangana   1059538765
6     west bengal   1030648875
7      tamil nadu    968325140
8         gujarat    903656540
9  madhya pradesh    903312230

### ⭐ Query 4 — Brand Market Share per State (Top 5 Brands)
#### ✔ SQL Query (Top 5 Brands per State)

In [31]:
query4 = """
SELECT states, brand, SUM(user_counts) AS total_users
FROM state_trans_user
GROUP BY states, brand
ORDER BY states, total_users DESC;
"""

brand_state = pd.read_sql(query4, conn)
brand_state.head()

states    brand  total_users
0  andaman and nicobar islands   Xiaomi       743760
1  andaman and nicobar islands     Vivo       705520
2  andaman and nicobar islands  Samsung       573110
3  andaman and nicobar islands     Oppo       272240
4  andaman and nicobar islands   Others       236765

#### ⭐ To extract Top 5 brands per state:

In [32]:
top5_brands_state = (
    brand_state
    .sort_values(['states', 'total_users'], ascending=[True, False])
    .groupby('states')
    .head(5)
    .reset_index(drop=True)
)

top5_brands_state.head(20)

states    brand  total_users
0   andaman and nicobar islands   Xiaomi       743760
1   andaman and nicobar islands     Vivo       705520
2   andaman and nicobar islands  Samsung       573110
3   andaman and nicobar islands     Oppo       272240
4   andaman and nicobar islands   Others       236765
5                andhra pradesh   Xiaomi    282899380
6                andhra pradesh  Samsung    218776440
7                andhra pradesh     Vivo    196386340
8                andhra pradesh     Oppo    122188630
9                andhra pradesh   Realme     98283940
10            arunachal pradesh   Xiaomi      3262820
11            arunachal pradesh     Vivo      3137840
12            arunachal pradesh  Samsung      2649890
13            arunachal pradesh     Oppo      2145610
14            arunachal pradesh   Realme      1104240
15                        assam   Xiaomi     50460230
16                        assam     Vivo     38845210
17                        assam  Samsung     28902575
18                        assam     Oppo     26707785
19                        assam   Others     20161840

### ⭐ Query 5 — Yearly Transaction Growth per State

In [34]:
pd.read_sql("""
SELECT 
    states,
    year,
    SUM(trans_counts) AS total_transactions,
    SUM(amount) AS total_amount
FROM state_trans_user
GROUP BY states, year
ORDER BY states, year;
""", conn)

states  year  total_transactions  total_amount
0    andaman and nicobar islands  2018              638231  2.079837e+09
1    andaman and nicobar islands  2019             1464144  5.208112e+09
2    andaman and nicobar islands  2020             4909014  1.426066e+10
3    andaman and nicobar islands  2021            18392583  5.418037e+10
4    andaman and nicobar islands  2022             8742228  2.341933e+10
..                           ...   ...                 ...           ...
175                  west bengal  2018          1320127985  1.227476e+12
176                  west bengal  2019          2836906490  2.926160e+12
177                  west bengal  2020          3346445597  6.080047e+12
178                  west bengal  2021          7335834407  1.481752e+13
179                  west bengal  2022          2865556012  6.065693e+12

[180 rows x 4 columns]

### ⭐ Query 6 — Year-over-Year (YoY) Growth of Transactions

In [35]:
pd.read_sql("""
WITH yearly AS (
    SELECT 
        states,
        year,
        SUM(trans_counts) AS total_txn
    FROM state_trans_user
    GROUP BY states, year
),
growth AS (
    SELECT 
        a.states,
        a.year,
        a.total_txn,
        (a.total_txn - b.total_txn) * 1.0 / b.total_txn AS yoy_growth
    FROM yearly a
    LEFT JOIN yearly b 
        ON a.states = b.states 
        AND a.year = b.year + 1
)
SELECT *
FROM growth
ORDER BY states, year;
""", conn)

states  year   total_txn  yoy_growth
0    andaman and nicobar islands  2018      638231         NaN
1    andaman and nicobar islands  2019     1464144    1.294066
2    andaman and nicobar islands  2020     4909014    2.352822
3    andaman and nicobar islands  2021    18392583    2.746696
4    andaman and nicobar islands  2022     8742228   -0.524687
..                           ...   ...         ...         ...
175                  west bengal  2018  1320127985         NaN
176                  west bengal  2019  2836906490    1.148963
177                  west bengal  2020  3346445597    0.179611
178                  west bengal  2021  7335834407    1.192127
179                  west bengal  2022  2865556012   -0.609376

[180 rows x 4 columns]

### ⭐ Query 7 — Total Transactions & Amount by Quarter (All India)

In [36]:
pd.read_sql("""
SELECT 
    year,
    quarter,
    SUM(trans_counts) AS total_transactions,
    SUM(amount) AS total_amount
FROM state_trans_user
GROUP BY year, quarter
ORDER BY year, quarter;
""", conn)

year  quarter  total_transactions  total_amount
0   2018        1          1478681589  1.890168e+12
1   2018        2          2061019840  3.348116e+12
2   2018        3          3754297404  5.226117e+12
3   2018        4          4588227677  7.389099e+12
4   2019        1          7798922791  1.089023e+13
5   2019        2          8969189856  1.489635e+13
6   2019        3         12045106425  1.839815e+13
7   2019        4         16064880293  2.485883e+13
8   2020        1         17853418506  2.966824e+13
9   2020        2         15934697790  2.910760e+13
10  2020        3         22363563013  4.292258e+13
11  2020        4         31562042842  5.935437e+13
12  2021        1         38033440093  6.912668e+13
13  2021        2         43355075115  8.224113e+13
14  2021        3         57901109970  1.013837e+14
15  2021        4         72883096242  1.278342e+14
16  2022        1         83524979560  1.456359e+14

#### ✅ Why values look like 1.890168e+12 ?

Pandas automatically formats very large numbers using scientific notation.

Example:

1.890168e+12 means
👉 1.890168 × 10¹²
👉 1,890,168,000,000 (≈ 1.89 trillion)

1.278342e+14 means
👉 1.278342 × 10¹⁴
👉 127,834,200,000,000 (≈ 127 trillion)

📌 Why Pandas does this?

Because:

total_amount is extremely large (billions/trillions).

Pandas switches to scientific notation for readability.

#### ✅ How to display full numbers instead of scientific notation?
Option 1 — Disable scientific notation for entire notebook
pd.set_option('display.float_format', '{:.2f}'.format)
This will show normal numbers like:

1890168000000.00
#### Option 2 — Convert amount to integer
(If values have no decimals)

df['total_amount'] = df['total_amount'].astype(int)

#### Option 3 — Format with commas
df['total_amount'] = df['total_amount'].apply(lambda x: f"{x:,.0f}")
Output example:

total_amount
1,890,168,000,000
127,834,200,000,000

### ✅ NEXT ANALYSIS STEP (Step 4): Identify the Fastest-Growing States (YoY Trend)

I will calculate YoY growth for each state and then find:

🔹 Top 10 fastest-growing states (2021 → 2022)
🔹 Top 10 fastest-growing states (2020 → 2021)
🔹 States with negative growth

This tells which states are exploding in digital payments and which are declining.

###### (creates df_yearly, computes YoY on both transactions and amount, and prints top/negative movers):

In [39]:
# build yearly agg from DB, compute YoY growth (transactions & amount)
import pandas as pd

df_yearly = pd.read_sql("""
SELECT 
    states,
    year,
    SUM(trans_counts) AS total_transactions,
    SUM(amount) AS total_amount
FROM state_trans_user
GROUP BY states, year
ORDER BY states, year;
""", conn)

# compute YoY pct change per state
df_yearly['txn_yoy'] = df_yearly.groupby('states')['total_transactions'].pct_change()
df_yearly['amt_yoy'] = df_yearly.groupby('states')['total_amount'].pct_change()

# Top 10 fastest-growing (transactions) for 2021 -> 2022
top_2022_txn = df_yearly[df_yearly['year'] == 2022].sort_values('txn_yoy', ascending=False).head(10)

# Top 10 fastest-growing (transactions) for 2020 -> 2021
top_2021_txn = df_yearly[df_yearly['year'] == 2021].sort_values('txn_yoy', ascending=False).head(10)

# States with negative growth (transactions) in 2021 -> 2022
neg_2022_txn = df_yearly[(df_yearly['year'] == 2022) & (df_yearly['txn_yoy'] < 0)].sort_values('txn_yoy')

# Also show top movers by amount for 2021->2022
top_2022_amt = df_yearly[df_yearly['year'] == 2022].sort_values('amt_yoy', ascending=False).head(10)

# Print results
print("Top 10 by txn YoY (2021->2022):")
display(top_2022_txn[['states','year','total_transactions','txn_yoy']])

print("\nTop 10 by txn YoY (2020->2021):")
display(top_2021_txn[['states','year','total_transactions','txn_yoy']])

print("\nStates with negative txn growth (2021->2022):")
display(neg_2022_txn[['states','year','total_transactions','txn_yoy']].head(20))

print("\nTop 10 by amount YoY (2021->2022):")
display(top_2022_amt[['states','year','total_amount','amt_yoy']])

Top 10 by txn YoY (2021->2022):


states  year  total_transactions  \
4                 andaman and nicobar islands  2022             8742228   
19                                      assam  2022           647788746   
49                                        goa  2022           134355078   
39   dadra and nagar haveli and daman and diu  2022            52830833   
34                               chhattisgarh  2022          1071546212   
74                                  jharkhand  2022          1220023310   
99                             madhya pradesh  2022          5199733638   
169                             uttar pradesh  2022          5213485706   
84                                     kerala  2022          1048090615   
24                                      bihar  2022          3128294774   

      txn_yoy  
4   -0.524687  
19  -0.526874  
49  -0.557431  
39  -0.576356  
34  -0.582063  
74  -0.582143  
99  -0.585735  
169 -0.591120  
84  -0.593441  
24  -0.595654


Top 10 by txn YoY (2020->2021):


states  year  total_transactions   txn_yoy
3    andaman and nicobar islands  2021            18392583  2.746696
68             jammu and kashmir  2021           516637638  2.675913
48                           goa  2021           303580266  2.039369
108                      manipur  2021           154205095  1.832113
133                   puducherry  2021           204064905  1.773740
98                madhya pradesh  2021         12551710556  1.730213
83                        kerala  2021          2577953202  1.691671
18                         assam  2021          1369167437  1.640695
23                         bihar  2021          7736675727  1.638519
103                  maharashtra  2021         31845905168  1.632691


States with negative txn growth (2021->2022):


states  year  total_transactions   txn_yoy
89              ladakh  2022             6930165 -0.667767
94         lakshadweep  2022              180367 -0.667464
119            mizoram  2022             6654087 -0.655470
69   jammu and kashmir  2022           178698421 -0.654113
9       andhra pradesh  2022          6408643021 -0.635143
164            tripura  2022            31583090 -0.634604
64    himachal pradesh  2022           176969980 -0.632933
29          chandigarh  2022            94605401 -0.625023
139             punjab  2022           635341322 -0.624544
124           nagaland  2022            21185472 -0.621989
109            manipur  2022            58450469 -0.620956
59             haryana  2022          2111874699 -0.620636
174        uttarakhand  2022           490168492 -0.615896
159          telangana  2022          9750987235 -0.615808
44               delhi  2022          3640998471 -0.614651
149             sikkim  2022            19187322 -0.614250
79           karnataka  2022         11636409477 -0.613846
179        west bengal  2022          2865556012 -0.609376
54             gujarat  2022          2316018606 -0.609267
154         tamil nadu  2022          3240304650 -0.605251


Top 10 by amount YoY (2021->2022):


states  year  total_amount   amt_yoy
19                                      assam  2022  1.455085e+12 -0.544496
4                 andaman and nicobar islands  2022  2.341933e+10 -0.567752
114                                 meghalaya  2022  5.128572e+10 -0.570960
49                                        goa  2022  2.179809e+11 -0.577256
179                               west bengal  2022  6.065693e+12 -0.590640
74                                  jharkhand  2022  2.141702e+12 -0.591043
34                               chhattisgarh  2022  1.736318e+12 -0.594374
169                             uttar pradesh  2022  9.714355e+12 -0.600294
129                                    odisha  2022  4.963500e+12 -0.607045
39   dadra and nagar haveli and daman and diu  2022  6.860753e+10 -0.610065

#### Check Columns in state_trans_user

In [43]:
pd.read_sql("PRAGMA table_info(state_trans_user);", conn)

cid          name     type  notnull dflt_value  pk
0    0        states     TEXT        0       None   0
1    1          year  INTEGER        0       None   0
2    2       quarter  INTEGER        0       None   0
3    3    trans_type     TEXT        0       None   0
4    4  trans_counts  INTEGER        0       None   0
5    5        amount     REAL        0       None   0
6    6         brand     TEXT        0       None   0
7    7   user_counts  INTEGER        0       None   0
8    8    percentage     REAL        0       None   0

#### Recreate the Required Table in SQLite

I will recreate:

state_yearly_transactions

columns:

states

year

total_transactions

total_amount

In [44]:
# STEP 1: Create state_yearly_transactions (corrected)
df_state_year = pd.read_sql("""
    SELECT 
        states,
        year,
        SUM(trans_counts) AS total_transactions,
        SUM(amount) AS total_amount
    FROM state_trans_user
    GROUP BY states, year
    ORDER BY states, year;
""", conn)

df_state_year.to_sql("state_yearly_transactions", conn, if_exists="replace", index=False)

print("✅ Table 'state_yearly_transactions' created successfully!")
df_state_year.head()

✅ Table 'state_yearly_transactions' created successfully!


states  year  total_transactions  total_amount
0  andaman and nicobar islands  2018              638231  2.079837e+09
1  andaman and nicobar islands  2019             1464144  5.208112e+09
2  andaman and nicobar islands  2020             4909014  1.426066e+10
3  andaman and nicobar islands  2021            18392583  5.418037e+10
4  andaman and nicobar islands  2022             8742228  2.341933e+10

### ⭐ Query 6 — Highest Transaction Amount State per Year

(From annual aggregated dataset: state_yearly_transactions)

In [45]:
query = """
SELECT year, states, total_amount
FROM (
    SELECT 
        year,
        states,
        total_amount,
        ROW_NUMBER() OVER (PARTITION BY year ORDER BY total_amount DESC) AS rn
    FROM state_yearly_transactions
)
WHERE rn = 1
ORDER BY year;
"""

df_top_amount_state = pd.read_sql(query, conn)
df_top_amount_state

year       states  total_amount
0  2018  maharashtra  2.088333e+12
1  2019    karnataka  8.703885e+12
2  2020    telangana  2.084706e+13
3  2021    telangana  4.916301e+13
4  2022  maharashtra  1.849539e+13

#### Query 7 — Highest Transaction Count State per Year

In [46]:
query = """
SELECT year, states, total_transactions
FROM (
    SELECT 
        year,
        states,
        total_transactions,
        ROW_NUMBER() OVER (PARTITION BY year ORDER BY total_transactions DESC) AS rn
    FROM state_yearly_transactions
)
WHERE rn = 1
ORDER BY year;
"""

df_top_txn_state = pd.read_sql(query, conn)
df_top_txn_state

year       states  total_transactions
0  2018  west bengal          1320127985
1  2019    karnataka          6715711200
2  2020    karnataka         13377684287
3  2021  maharashtra         31845905168
4  2022  maharashtra         12864570818

#### Query 8 — Top 5 Districts by Transaction Amount per State

In [47]:
query = """
SELECT states, district, total_amount
FROM (
    SELECT 
        states,
        district,
        SUM(amount) AS total_amount,
        ROW_NUMBER() OVER (PARTITION BY states ORDER BY SUM(amount) DESC) AS rn
    FROM district_transactions
    GROUP BY states, district
)
WHERE rn <= 5
ORDER BY states, total_amount DESC;
"""

df_top5_districts = pd.read_sql(query, conn)
df_top5_districts

states                             district  \
0    andaman and nicobar islands               south andaman district   
1    andaman and nicobar islands    north and middle andaman district   
2    andaman and nicobar islands                    nicobars district   
3                 andhra pradesh               visakhapatnam district   
4                 andhra pradesh                     krishna district   
..                           ...                                  ...   
155                  west bengal                     kolkata district   
156                  west bengal  north twenty four parganas district   
157                  west bengal                 murshidabad district   
158                  west bengal                      howrah district   
159                  west bengal             purba medinipur district   

     total_amount  
0    1.666020e+10  
1    1.448192e+09  
2    8.863123e+08  
3    1.609891e+12  
4    1.239996e+12  
..            ...  
155  1.495338e+12  
156  5.506193e+11  
157  3.172991e+11  
158  3.014211e+11  
159  2.425113e+11  

[160 rows x 3 columns]

#### Query 9 — Top 5 Districts by Transaction Count per State

In [48]:
query = """
SELECT states, district, total_transactions
FROM (
    SELECT 
        states,
        district,
        SUM(trans_counts) AS total_transactions,
        ROW_NUMBER() OVER (PARTITION BY states ORDER BY SUM(trans_counts) DESC) AS rn
    FROM district_transactions
    GROUP BY states, district
)
WHERE rn <= 5
ORDER BY states, total_transactions DESC;
"""

df_top5_districts_txn = pd.read_sql(query, conn)
df_top5_districts_txn

states                             district  \
0    andaman and nicobar islands               south andaman district   
1    andaman and nicobar islands    north and middle andaman district   
2    andaman and nicobar islands                    nicobars district   
3                 andhra pradesh               visakhapatnam district   
4                 andhra pradesh                     krishna district   
..                           ...                                  ...   
155                  west bengal                     kolkata district   
156                  west bengal  north twenty four parganas district   
157                  west bengal                      howrah district   
158                  west bengal                 murshidabad district   
159                  west bengal  south twenty four parganas district   

     total_transactions  
0               6907007  
1                554604  
2                285720  
3             851575387  
4             560241914  
..                  ...  
155           788173142  
156           331646674  
157           165085079  
158           146568619  
159           117756703  

[160 rows x 3 columns]

#### Query 10 — Top 5 Brands by Users per State

In [49]:
query = """
SELECT states, brand, total_users
FROM (
    SELECT 
        states,
        brand,
        SUM(user_counts) AS total_users,
        ROW_NUMBER() OVER (PARTITION BY states ORDER BY SUM(user_counts) DESC) AS rn
    FROM state_trans_user
    GROUP BY states, brand
)
WHERE rn <= 5
ORDER BY states, total_users DESC;
"""

df_top5_brands = pd.read_sql(query, conn)
df_top5_brands

states    brand  total_users
0    andaman and nicobar islands   Xiaomi       743760
1    andaman and nicobar islands     Vivo       705520
2    andaman and nicobar islands  Samsung       573110
3    andaman and nicobar islands     Oppo       272240
4    andaman and nicobar islands   Others       236765
..                           ...      ...          ...
175                  west bengal   Xiaomi    270827230
176                  west bengal  Samsung    213833040
177                  west bengal     Vivo    143989870
178                  west bengal   Others    119279310
179                  west bengal     Oppo    107965060

[180 rows x 3 columns]

#### Query 11 — YoY Growth of Transactions per State

In [50]:
# Read state_yearly_transactions table
df_state_year = pd.read_sql("SELECT * FROM state_yearly_transactions;", conn)

# Calculate YoY growth
df_state_year['txn_yoy'] = df_state_year.groupby('states')['total_transactions'].pct_change()
df_state_year['amt_yoy'] = df_state_year.groupby('states')['total_amount'].pct_change()

df_state_year.head(10)

states  year  total_transactions  total_amount  \
0  andaman and nicobar islands  2018              638231  2.079837e+09   
1  andaman and nicobar islands  2019             1464144  5.208112e+09   
2  andaman and nicobar islands  2020             4909014  1.426066e+10   
3  andaman and nicobar islands  2021            18392583  5.418037e+10   
4  andaman and nicobar islands  2022             8742228  2.341933e+10   
5               andhra pradesh  2018           855570232  1.342792e+12   
6               andhra pradesh  2019          3268982794  5.915705e+12   
7               andhra pradesh  2020          8399783689  1.844971e+13   
8               andhra pradesh  2021         17564818227  4.030031e+13   
9               andhra pradesh  2022          6408643021  1.423753e+13   

    txn_yoy   amt_yoy  
0       NaN       NaN  
1  1.294066  1.504096  
2  2.352822  1.738162  
3  2.746696  2.799290  
4 -0.524687 -0.567752  
5       NaN       NaN  
6  2.820823  3.405526  
7  1.569541  2.118768  
8  1.091104  1.184333  
9 -0.635143 -0.646714

#### Query 12 — Top 10 States by Transaction YoY (Latest Year)

In [51]:
# Filter for latest year (2022) and sort by txn_yoy descending
df_top10_txn_yoy = df_state_year[df_state_year['year'] == 2022].sort_values(by='txn_yoy', ascending=False).head(10)
df_top10_txn_yoy[['states','year','total_transactions','txn_yoy']]

states  year  total_transactions  \
4                 andaman and nicobar islands  2022             8742228   
19                                      assam  2022           647788746   
49                                        goa  2022           134355078   
39   dadra and nagar haveli and daman and diu  2022            52830833   
34                               chhattisgarh  2022          1071546212   
74                                  jharkhand  2022          1220023310   
99                             madhya pradesh  2022          5199733638   
169                             uttar pradesh  2022          5213485706   
84                                     kerala  2022          1048090615   
24                                      bihar  2022          3128294774   

      txn_yoy  
4   -0.524687  
19  -0.526874  
49  -0.557431  
39  -0.576356  
34  -0.582063  
74  -0.582143  
99  -0.585735  
169 -0.591120  
84  -0.593441  
24  -0.595654

#### Query 13 — Top 10 States by Amount YoY (Latest Year)

In [52]:
df_top10_amt_yoy = df_state_year[df_state_year['year'] == 2022].sort_values(by='amt_yoy', ascending=False).head(10)
df_top10_amt_yoy[['states','year','total_amount','amt_yoy']]

states  year  total_amount   amt_yoy
19                                      assam  2022  1.455085e+12 -0.544496
4                 andaman and nicobar islands  2022  2.341933e+10 -0.567752
114                                 meghalaya  2022  5.128572e+10 -0.570960
49                                        goa  2022  2.179809e+11 -0.577256
179                               west bengal  2022  6.065693e+12 -0.590640
74                                  jharkhand  2022  2.141702e+12 -0.591043
34                               chhattisgarh  2022  1.736318e+12 -0.594374
169                             uttar pradesh  2022  9.714355e+12 -0.600294
129                                    odisha  2022  4.963500e+12 -0.607045
39   dadra and nagar haveli and daman and diu  2022  6.860753e+10 -0.610065

#### Query 14 — States with Negative Transaction Growth (Latest Year)

In [53]:
df_negative_txn = df_state_year[(df_state_year['year'] == 2022) & (df_state_year['txn_yoy'] < 0)].sort_values(by='txn_yoy')
df_negative_txn[['states','year','total_transactions','txn_yoy']]

states  year  total_transactions  \
89                                     ladakh  2022             6930165   
94                                lakshadweep  2022              180367   
119                                   mizoram  2022             6654087   
69                          jammu and kashmir  2022           178698421   
9                              andhra pradesh  2022          6408643021   
164                                   tripura  2022            31583090   
64                           himachal pradesh  2022           176969980   
29                                 chandigarh  2022            94605401   
139                                    punjab  2022           635341322   
124                                  nagaland  2022            21185472   
109                                   manipur  2022            58450469   
59                                    haryana  2022          2111874699   
174                               uttarakhand  2022           490168492   
159                                 telangana  2022          9750987235   
44                                      delhi  2022          3640998471   
149                                    sikkim  2022            19187322   
79                                  karnataka  2022         11636409477   
179                               west bengal  2022          2865556012   
54                                    gujarat  2022          2316018606   
154                                tamil nadu  2022          3240304650   
144                                 rajasthan  2022          6214584420   
134                                puducherry  2022            80950936   
14                          arunachal pradesh  2022            30844748   
114                                 meghalaya  2022            18422250   
129                                    odisha  2022          2899968489   
104                               maharashtra  2022         12864570818   
24                                      bihar  2022          3128294774   
84                                     kerala  2022          1048090615   
169                             uttar pradesh  2022          5213485706   
99                             madhya pradesh  2022          5199733638   
74                                  jharkhand  2022          1220023310   
34                               chhattisgarh  2022          1071546212   
39   dadra and nagar haveli and daman and diu  2022            52830833   
49                                        goa  2022           134355078   
19                                      assam  2022           647788746   
4                 andaman and nicobar islands  2022             8742228   

      txn_yoy  
89  -0.667767  
94  -0.667464  
119 -0.655470  
69  -0.654113  
9   -0.635143  
164 -0.634604  
64  -0.632933  
29  -0.625023  
139 -0.624544  
124 -0.621989  
109 -0.620956  
59  -0.620636  
174 -0.615896  
159 -0.615808  
44  -0.614651  
149 -0.614250  
79  -0.613846  
179 -0.609376  
54  -0.609267  
154 -0.605251  
144 -0.604400  
134 -0.603308  
14  -0.601885  
114 -0.600790  
129 -0.597243  
104 -0.596037  
24  -0.595654  
84  -0.593441  
169 -0.591120  
99  -0.585735  
74  -0.582143  
34  -0.582063  
39  -0.576356  
49  -0.557431  
19  -0.526874  
4   -0.524687

## ✅Next logical step is district-level YoY analysis

#### Query 15 — District-wise Total Transactions per Year

In [54]:
query_district_txn = """
SELECT states, district, year, 
       SUM(trans_counts) AS total_transactions
FROM district_transactions
GROUP BY states, district, year
ORDER BY states, district, year;
"""
df_district_txn = pd.read_sql(query_district_txn, conn)
df_district_txn.head()

states           district  year  total_transactions
0  andaman and nicobar islands  nicobars district  2018                4604
1  andaman and nicobar islands  nicobars district  2019                8267
2  andaman and nicobar islands  nicobars district  2020               20907
3  andaman and nicobar islands  nicobars district  2021               70800
4  andaman and nicobar islands  nicobars district  2022              181142

#### Query 16 — District-wise Total Amount per Year

In [55]:
query_district_amt = """
SELECT states, district, year, 
       SUM(amount) AS total_amount
FROM district_transactions
GROUP BY states, district, year
ORDER BY states, district, year;
"""
df_district_amt = pd.read_sql(query_district_amt, conn)
df_district_amt.head()

states           district  year  total_amount
0  andaman and nicobar islands  nicobars district  2018  1.778097e+07
1  andaman and nicobar islands  nicobars district  2019  4.029829e+07
2  andaman and nicobar islands  nicobars district  2020  8.382638e+07
3  andaman and nicobar islands  nicobars district  2021  2.434057e+08
4  andaman and nicobar islands  nicobars district  2022  5.010010e+08

#### Query 17 — Top 5 Districts by Transactions (Latest Year)

In [56]:
query_top_district_txn = """
SELECT states, district, SUM(trans_counts) AS total_transactions
FROM district_transactions
WHERE year = 2022
GROUP BY states, district
ORDER BY total_transactions DESC
LIMIT 5;
"""
df_top_district_txn = pd.read_sql(query_top_district_txn, conn)
df_top_district_txn

states                  district  total_transactions
0    karnataka  bengaluru urban district          3050041216
1    telangana        hyderabad district          2971669634
2  maharashtra             pune district          1994176959
3    rajasthan           jaipur district          1359854306
4        delhi          central district           811445152

#### Query 18 — Top 5 Districts by Transaction Amount (Latest Year)

In [57]:
query_top_district_amt = """
SELECT states, district, SUM(amount) AS total_amount
FROM district_transactions
WHERE year = 2022
GROUP BY states, district
ORDER BY total_amount DESC
LIMIT 5;
"""
df_top_district_amt = pd.read_sql(query_top_district_amt, conn)
df_top_district_amt

states                  district  total_amount
0    telangana        hyderabad district  4.777268e+12
1    karnataka  bengaluru urban district  3.874838e+12
2  maharashtra             pune district  2.269088e+12
3    rajasthan           jaipur district  2.132277e+12
4        delhi          central district  1.144284e+12

#### Query 19 — Top 5 States by Registered Users (Latest Year)

In [59]:
query_top_states_users = """
SELECT states, SUM(user_counts) AS total_users
FROM state_trans_user
WHERE year = 2022
GROUP BY states
ORDER BY total_users DESC
LIMIT 5;
"""
df_top_states_users = pd.read_sql(query_top_states_users, conn)
df_top_states_users

states  total_users
0     maharashtra    241138150
1   uttar pradesh    194998460
2       karnataka    156001745
3  andhra pradesh    119289755
4       rajasthan    116994285

#### Query 20 — Top 5 Districts by Registered Users (Latest Year)

In [60]:
query_top_district_users = """
SELECT states, district, SUM(registered_user_counts) AS total_users
FROM district_transactions
WHERE year = 2022
GROUP BY states, district
ORDER BY total_users DESC
LIMIT 5;
"""
df_top_district_users = pd.read_sql(query_top_district_users, conn)
df_top_district_users

states                  district  total_users
0    karnataka  bengaluru urban district     54702637
1  maharashtra             pune district     35905243
2  maharashtra            thane district     22467231
3    rajasthan           jaipur district     22215879
4  maharashtra  mumbai suburban district     21685243

#### Query 21 — Transaction Count Share by Type (Latest Year, All States)

In [61]:
query_txn_type_share = """
SELECT trans_type, SUM(trans_counts) AS total_transactions
FROM state_trans_user
WHERE year = 2022
GROUP BY trans_type
ORDER BY total_transactions DESC;
"""
df_txn_type_share = pd.read_sql(query_txn_type_share, conn)
df_txn_type_share

trans_type  total_transactions
0         Merchant payments         41237254344
1     Peer-to-peer payments         33816629418
2  Recharge & bill payments          8246382463
3                    Others           180327697
4        Financial Services            44385638

#### Query 22 — Transaction Amount Share by Type (Latest Year, All States)

In [62]:
query_txn_amount_share = """
SELECT trans_type, SUM(amount) AS total_amount
FROM state_trans_user
WHERE year = 2022
GROUP BY trans_type
ORDER BY total_amount DESC;
"""
df_txn_amount_share = pd.read_sql(query_txn_amount_share, conn)
df_txn_amount_share

trans_type  total_amount
0     Peer-to-peer payments  1.155647e+14
1         Merchant payments  2.500087e+13
2  Recharge & bill payments  4.923578e+12
3                    Others  9.264926e+10
4        Financial Services  5.413275e+10

#### Query 23 — Top 5 States by “Recharge & Bill Payments” Amount (Latest Year)

In [63]:
query_recharge_top_states = """
SELECT states, SUM(amount) AS total_amount
FROM state_trans_user
WHERE year = 2022 AND trans_type = 'Recharge & bill payments'
GROUP BY states
ORDER BY total_amount DESC
LIMIT 5;
"""
df_recharge_top_states = pd.read_sql(query_recharge_top_states, conn)
df_recharge_top_states

states  total_amount
0     maharashtra  7.095865e+11
1       telangana  5.551750e+11
2       karnataka  4.828044e+11
3       rajasthan  3.909717e+11
4  andhra pradesh  3.792034e+11

#### Query 24 — Top 5 States by “Peer-to-Peer Payments” Amount (Latest Year)

In [64]:
query_p2p_top_states = """
SELECT states, SUM(amount) AS total_amount
FROM state_trans_user
WHERE year = 2022 AND trans_type = 'Peer-to-peer payments'
GROUP BY states
ORDER BY total_amount DESC
LIMIT 5;
"""
df_p2p_top_states = pd.read_sql(query_p2p_top_states, conn)
df_p2p_top_states

states  total_amount
0       telangana  1.512819e+13
1     maharashtra  1.402378e+13
2       karnataka  1.367519e+13
3  andhra pradesh  1.198592e+13
4       rajasthan  8.959780e+12

#### Query 25 — Top 5 Districts by Transaction Amount

In [70]:
query_top5_districts_amount = """
-- Query 25: Top 5 Districts by Transaction Amount (Latest Year)
SELECT states, district, SUM(amount) AS total_amount
FROM district_transactions
WHERE year = 2022
GROUP BY states, district
ORDER BY total_amount DESC
LIMIT 5;
"""

df_top5_districts_amount = pd.read_sql(query_top5_districts_amount, conn)
df_top5_districts_amount

states                  district  total_amount
0    telangana        hyderabad district  4.777268e+12
1    karnataka  bengaluru urban district  3.874838e+12
2  maharashtra             pune district  2.269088e+12
3    rajasthan           jaipur district  2.132277e+12
4        delhi          central district  1.144284e+12

In [68]:
pd.read_sql("PRAGMA table_info(district_transactions);", conn)

cid                    name     type  notnull dflt_value  pk
0    0                  states     TEXT        0       None   0
1    1                    year  INTEGER        0       None   0
2    2                 quarter  INTEGER        0       None   0
3    3                district     TEXT        0       None   0
4    4            trans_counts  INTEGER        0       None   0
5    5                  amount     REAL        0       None   0
6    6  registered_user_counts  INTEGER        0       None   0

#### Query 26 — Top 5 Districts by Transaction Count (Latest Year):

In [71]:
query_top5_districts_count = """
-- Query 26: Top 5 Districts by Transaction Count (Latest Year)
SELECT states, district, SUM(trans_counts) AS total_transactions
FROM district_transactions
WHERE year = 2022
GROUP BY states, district
ORDER BY total_transactions DESC
LIMIT 5;
"""

df_top5_districts_count = pd.read_sql(query_top5_districts_count, conn)
df_top5_districts_count

states                  district  total_transactions
0    karnataka  bengaluru urban district          3050041216
1    telangana        hyderabad district          2971669634
2  maharashtra             pune district          1994176959
3    rajasthan           jaipur district          1359854306
4        delhi          central district           811445152

#### Query 27 — Top 5 States by Total Users (Latest Year):

In [72]:
query_top5_states_users = """
-- Query 27: Top 5 States by Total Users (Latest Year)
SELECT states, SUM(user_counts) AS total_users
FROM state_trans_user
WHERE year = 2022
GROUP BY states
ORDER BY total_users DESC
LIMIT 5;
"""

df_top5_states_users = pd.read_sql(query_top5_states_users, conn)
df_top5_states_users

states  total_users
0     maharashtra    241138150
1   uttar pradesh    194998460
2       karnataka    156001745
3  andhra pradesh    119289755
4       rajasthan    116994285

#### Query 28 — Top 5 Districts by Total Users (Latest Year):

In [73]:
query_top5_districts_users = """
-- Query 28: Top 5 Districts by Total Users (Latest Year)
SELECT states, district, SUM(registered_user_counts) AS total_users
FROM district_transactions
WHERE year = 2022
GROUP BY states, district
ORDER BY total_users DESC
LIMIT 5;
"""

df_top5_districts_users = pd.read_sql(query_top5_districts_users, conn)
df_top5_districts_users

states                  district  total_users
0    karnataka  bengaluru urban district     54702637
1  maharashtra             pune district     35905243
2  maharashtra            thane district     22467231
3    rajasthan           jaipur district     22215879
4  maharashtra  mumbai suburban district     21685243

#### Query 29 — Top 5 States by “Recharge & Bill Payments” Amount (Latest Year):

In [74]:
query_top5_states_recharge = """
-- Query 29: Top 5 States by "Recharge & Bill Payments" Amount (Latest Year)
SELECT states, SUM(amount) AS total_amount
FROM state_trans_user
WHERE year = 2022 AND trans_type = 'Recharge & bill payments'
GROUP BY states
ORDER BY total_amount DESC
LIMIT 5;
"""

df_top5_states_recharge = pd.read_sql(query_top5_states_recharge, conn)
df_top5_states_recharge

states  total_amount
0     maharashtra  7.095865e+11
1       telangana  5.551750e+11
2       karnataka  4.828044e+11
3       rajasthan  3.909717e+11
4  andhra pradesh  3.792034e+11

#### Query 30 — Top 5 States by “Peer-to-Peer Payments” Amount (Latest Year):

In [75]:
query_top5_states_p2p = """
-- Query 30: Top 5 States by "Peer-to-Peer Payments" Amount (Latest Year)
SELECT states, SUM(amount) AS total_amount
FROM state_trans_user
WHERE year = 2022 AND trans_type = 'Peer-to-peer payments'
GROUP BY states
ORDER BY total_amount DESC
LIMIT 5;
"""

df_top5_states_p2p = pd.read_sql(query_top5_states_p2p, conn)
df_top5_states_p2p

states  total_amount
0       telangana  1.512819e+13
1     maharashtra  1.402378e+13
2       karnataka  1.367519e+13
3  andhra pradesh  1.198592e+13
4       rajasthan  8.959780e+12

#### Query 31 — Top 5 Districts by Transaction Amount (Latest Year, All Types):

In [76]:
query_top5_districts_amount = """
-- Query 31: Top 5 Districts by Transaction Amount (Latest Year, all transaction types)
SELECT states, district, SUM(amount) AS total_amount
FROM district_transactions
WHERE year = 2022
GROUP BY states, district
ORDER BY total_amount DESC
LIMIT 5;
"""

df_top5_districts_amount = pd.read_sql(query_top5_districts_amount, conn)
df_top5_districts_amount

states                  district  total_amount
0    telangana        hyderabad district  4.777268e+12
1    karnataka  bengaluru urban district  3.874838e+12
2  maharashtra             pune district  2.269088e+12
3    rajasthan           jaipur district  2.132277e+12
4        delhi          central district  1.144284e+12

#### Query 32 — Top 5 Districts by Transaction Count (Latest Year, All Types):

In [77]:
query_top5_districts_count = """
-- Query 32: Top 5 Districts by Transaction Count (Latest Year, all transaction types)
SELECT states, district, SUM(trans_counts) AS total_transactions
FROM district_transactions
WHERE year = 2022
GROUP BY states, district
ORDER BY total_transactions DESC
LIMIT 5;
"""

df_top5_districts_count = pd.read_sql(query_top5_districts_count, conn)
df_top5_districts_count

states                  district  total_transactions
0    karnataka  bengaluru urban district          3050041216
1    telangana        hyderabad district          2971669634
2  maharashtra             pune district          1994176959
3    rajasthan           jaipur district          1359854306
4        delhi          central district           811445152

#### Query 33 — Top 5 States by Total Users (Latest Year):

In [78]:
query_top5_states_users = """
-- Query 33: Top 5 States by Total Users (Latest Year)
SELECT states, SUM(user_counts) AS total_users
FROM state_trans_user
WHERE year = 2022
GROUP BY states
ORDER BY total_users DESC
LIMIT 5;
"""

df_top5_states_users = pd.read_sql(query_top5_states_users, conn)
df_top5_states_users

states  total_users
0     maharashtra    241138150
1   uttar pradesh    194998460
2       karnataka    156001745
3  andhra pradesh    119289755
4       rajasthan    116994285

#### Query 34 — Top 5 Districts by Total Users (Latest Year):

In [79]:
query_top5_districts_users = """
-- Query 34: Top 5 Districts by Total Users (Latest Year)
SELECT states, district, SUM(registered_user_counts) AS total_users
FROM district_transactions
WHERE year = 2022
GROUP BY states, district
ORDER BY total_users DESC
LIMIT 5;
"""

df_top5_districts_users = pd.read_sql(query_top5_districts_users, conn)
df_top5_districts_users

states                  district  total_users
0    karnataka  bengaluru urban district     54702637
1  maharashtra             pune district     35905243
2  maharashtra            thane district     22467231
3    rajasthan           jaipur district     22215879
4  maharashtra  mumbai suburban district     21685243

#### Query 35 — Quarterly Transaction & Amount per State

In [82]:
query_quarterly_state = """
SELECT 
    states,
    year,
    quarter,
    SUM(trans_counts) AS total_transactions,
    SUM(amount) AS total_amount
FROM state_trans_user
GROUP BY states, year, quarter
ORDER BY states, year, quarter;
"""
df_quarterly_state = pd.read_sql(query_quarterly_state, conn)
df_quarterly_state.head()

states  year  quarter  total_transactions  \
0  andaman and nicobar islands  2018        1               73238   
1  andaman and nicobar islands  2018        2              124740   
2  andaman and nicobar islands  2018        3              178915   
3  andaman and nicobar islands  2018        4              261338   
4  andaman and nicobar islands  2019        1              335346   

   total_amount  
0  1.609494e+08  
1  3.117239e+08  
2  6.111322e+08  
3  9.960317e+08  
4  1.125297e+09

#### Query 36 — Quarterly Transaction & Amount per District

In [83]:
query_quarterly_district = """
SELECT 
    states,
    district,
    year,
    quarter,
    SUM(trans_counts) AS total_transactions,
    SUM(amount) AS total_amount
FROM district_transactions
GROUP BY states, district, year, quarter
ORDER BY states, district, year, quarter;
"""
df_quarterly_district = pd.read_sql(query_quarterly_district, conn)
df_quarterly_district.head()

states           district  year  quarter  \
0  andaman and nicobar islands  nicobars district  2018        1   
1  andaman and nicobar islands  nicobars district  2018        2   
2  andaman and nicobar islands  nicobars district  2018        3   
3  andaman and nicobar islands  nicobars district  2018        4   
4  andaman and nicobar islands  nicobars district  2019        1   

   total_transactions  total_amount  
0                 528  1.139849e+06  
1                1120  3.072437e+06  
2                1471  6.387829e+06  
3                1485  7.180859e+06  
4                1539  7.236060e+06

#### Query 37 — Quarterly Top 5 Districts by Transaction Amount (Latest Year)

In [84]:
query_quarterly_top_districts = """
SELECT 
    states,
    district,
    SUM(amount) AS total_amount
FROM district_transactions
WHERE year = 2022
GROUP BY states, district
ORDER BY total_amount DESC
LIMIT 5;
"""
df_quarterly_top_districts = pd.read_sql(query_quarterly_top_districts, conn)
df_quarterly_top_districts

states                  district  total_amount
0    telangana        hyderabad district  4.777268e+12
1    karnataka  bengaluru urban district  3.874838e+12
2  maharashtra             pune district  2.269088e+12
3    rajasthan           jaipur district  2.132277e+12
4        delhi          central district  1.144284e+12

#### Query 38 — Quarterly Top 5 States by Transaction Amount (Latest Year)

In [85]:
query_quarterly_top_states = """
SELECT 
    states,
    SUM(amount) AS total_amount
FROM state_trans_user
WHERE year = 2022
GROUP BY states
ORDER BY total_amount DESC
LIMIT 5;
"""
df_quarterly_top_states = pd.read_sql(query_quarterly_top_states, conn)
df_quarterly_top_states

states  total_amount
0     maharashtra  1.849539e+13
1       telangana  1.837702e+13
2       karnataka  1.723314e+13
3  andhra pradesh  1.423753e+13
4       rajasthan  1.124255e+13

#### Query 39 — Quarterly Top 5 Districts by Transaction Amount (Latest Year)

In [86]:
query_quarterly_top_districts = """
SELECT 
    states,
    district,
    SUM(amount) AS total_amount
FROM district_transactions
WHERE year = 2022
GROUP BY states, district
ORDER BY total_amount DESC
LIMIT 5;
"""
df_quarterly_top_districts = pd.read_sql(query_quarterly_top_districts, conn)
df_quarterly_top_districts

states                  district  total_amount
0    telangana        hyderabad district  4.777268e+12
1    karnataka  bengaluru urban district  3.874838e+12
2  maharashtra             pune district  2.269088e+12
3    rajasthan           jaipur district  2.132277e+12
4        delhi          central district  1.144284e+12

#### Query 40 — Top 5 States by Transaction Amount (Latest Year, All Types)

In [87]:
query_top_states_amount_latest = """
SELECT 
    states,
    SUM(amount) AS total_amount
FROM state_trans_user
WHERE year = 2022
GROUP BY states
ORDER BY total_amount DESC
LIMIT 5;
"""
df_top_states_amount_latest = pd.read_sql(query_top_states_amount_latest, conn)
df_top_states_amount_latest

states  total_amount
0     maharashtra  1.849539e+13
1       telangana  1.837702e+13
2       karnataka  1.723314e+13
3  andhra pradesh  1.423753e+13
4       rajasthan  1.124255e+13

#### Query 41 — Top 5 Districts by Transaction Amount (Latest Year, All Types)

In [88]:
query_top_districts_amount_latest = """
SELECT 
    states,
    district,
    SUM(amount) AS total_amount
FROM district_transactions
WHERE year = 2022
GROUP BY states, district
ORDER BY total_amount DESC
LIMIT 5;
"""
df_top_districts_amount_latest = pd.read_sql(query_top_districts_amount_latest, conn)
df_top_districts_amount_latest

states                  district  total_amount
0    telangana        hyderabad district  4.777268e+12
1    karnataka  bengaluru urban district  3.874838e+12
2  maharashtra             pune district  2.269088e+12
3    rajasthan           jaipur district  2.132277e+12
4        delhi          central district  1.144284e+12

#### Query 42 — Top 5 States by Transaction Amount (Latest Year, All Types)

In [89]:
query_top_states_amount_latest = """
SELECT 
    states,
    SUM(amount) AS total_amount
FROM state_trans_user
WHERE year = 2022
GROUP BY states
ORDER BY total_amount DESC
LIMIT 5;
"""
df_top_states_amount_latest = pd.read_sql(query_top_states_amount_latest, conn)
df_top_states_amount_latest

states  total_amount
0     maharashtra  1.849539e+13
1       telangana  1.837702e+13
2       karnataka  1.723314e+13
3  andhra pradesh  1.423753e+13
4       rajasthan  1.124255e+13

#### Query 43 — Top 5 Districts by Transaction Amount (Latest Year, All Types)

In [90]:
query_top_districts_amount_latest = """
SELECT 
    states,
    district,
    SUM(amount) AS total_amount
FROM district_transactions
WHERE year = 2022
GROUP BY states, district
ORDER BY total_amount DESC
LIMIT 5;
"""
df_top_districts_amount_latest = pd.read_sql(query_top_districts_amount_latest, conn)
df_top_districts_amount_latest

states                  district  total_amount
0    telangana        hyderabad district  4.777268e+12
1    karnataka  bengaluru urban district  3.874838e+12
2  maharashtra             pune district  2.269088e+12
3    rajasthan           jaipur district  2.132277e+12
4        delhi          central district  1.144284e+12

#### Query 44 — Top 5 States by Transaction Amount (Latest Year, All Types)

In [91]:
query_top_states_amount_latest = """
SELECT 
    states,
    SUM(amount) AS total_amount
FROM state_trans_user
WHERE year = 2022
GROUP BY states
ORDER BY total_amount DESC
LIMIT 5;
"""
df_top_states_amount_latest = pd.read_sql(query_top_states_amount_latest, conn)
df_top_states_amount_latest

states  total_amount
0     maharashtra  1.849539e+13
1       telangana  1.837702e+13
2       karnataka  1.723314e+13
3  andhra pradesh  1.423753e+13
4       rajasthan  1.124255e+13

#### Query 45 — Top 5 Districts by Transaction Amount (Latest Year, All Types)

In [92]:
query_top_districts_amount_latest = """
SELECT 
    district,
    states,
    SUM(amount) AS total_amount
FROM district_transactions
WHERE year = 2022
GROUP BY states, district
ORDER BY total_amount DESC
LIMIT 5;
"""
df_top_districts_amount_latest = pd.read_sql(query_top_districts_amount_latest, conn)
df_top_districts_amount_latest

district       states  total_amount
0        hyderabad district    telangana  4.777268e+12
1  bengaluru urban district    karnataka  3.874838e+12
2             pune district  maharashtra  2.269088e+12
3           jaipur district    rajasthan  2.132277e+12
4          central district        delhi  1.144284e+12

#### Query 46 — Top 5 States by Transaction Amount (Latest Year)

In [93]:
query_top_states_amount_latest = """
SELECT 
    states,
    SUM(amount) AS total_amount
FROM state_trans_user
WHERE year = 2022
GROUP BY states
ORDER BY total_amount DESC
LIMIT 5;
"""
df_top_states_amount_latest = pd.read_sql(query_top_states_amount_latest, conn)
df_top_states_amount_latest

states  total_amount
0     maharashtra  1.849539e+13
1       telangana  1.837702e+13
2       karnataka  1.723314e+13
3  andhra pradesh  1.423753e+13
4       rajasthan  1.124255e+13

#### Query 47 — Top 5 Districts by Transaction Amount (Latest Year)

In [94]:
query_top_districts_amount_latest = """
SELECT 
    district,
    states,
    SUM(amount) AS total_amount
FROM district_transactions
WHERE year = 2022
GROUP BY district, states
ORDER BY total_amount DESC
LIMIT 5;
"""
df_top_districts_amount_latest = pd.read_sql(query_top_districts_amount_latest, conn)
df_top_districts_amount_latest

district       states  total_amount
0        hyderabad district    telangana  4.777268e+12
1  bengaluru urban district    karnataka  3.874838e+12
2             pune district  maharashtra  2.269088e+12
3           jaipur district    rajasthan  2.132277e+12
4          central district        delhi  1.144284e+12

#### Query 48 — Top 5 States by Transaction Amount (Latest Year)

In [95]:
query_top_states_amount_latest = """
SELECT 
    states,
    SUM(amount) AS total_amount
FROM state_trans_user
WHERE year = 2022
GROUP BY states
ORDER BY total_amount DESC
LIMIT 5;
"""
df_top_states_amount_latest = pd.read_sql(query_top_states_amount_latest, conn)
df_top_states_amount_latest

states  total_amount
0     maharashtra  1.849539e+13
1       telangana  1.837702e+13
2       karnataka  1.723314e+13
3  andhra pradesh  1.423753e+13
4       rajasthan  1.124255e+13

In [98]:
pd.read_sql("PRAGMA table_info(district_transactions);", conn)

cid                    name     type  notnull dflt_value  pk
0    0                  states     TEXT        0       None   0
1    1                    year  INTEGER        0       None   0
2    2                 quarter  INTEGER        0       None   0
3    3                district     TEXT        0       None   0
4    4            trans_counts  INTEGER        0       None   0
5    5                  amount     REAL        0       None   0
6    6  registered_user_counts  INTEGER        0       None   0

#### Query 49 — Top 5 Districts by Transaction Amount (Latest Year)

In [100]:
query_top_districts_amount_latest = """
SELECT 
states,
district,
SUM(amount) AS total_amount
FROM district_transactions
WHERE year = 2022
GROUP BY states, district
ORDER BY total_amount DESC
LIMIT 5;
"""

df_top_districts_amount_latest = pd.read_sql(query_top_districts_amount_latest, conn)
df_top_districts_amount_latest

states                  district  total_amount
0    telangana        hyderabad district  4.777268e+12
1    karnataka  bengaluru urban district  3.874838e+12
2  maharashtra             pune district  2.269088e+12
3    rajasthan           jaipur district  2.132277e+12
4        delhi          central district  1.144284e+12

#### Query50-- top 5 districts by transaction count

In [101]:
query_top_districts_txn_latest = """
SELECT 
states,
district,
SUM(trans_counts) AS total_transactions
FROM district_transactions
WHERE year = 2022
GROUP BY states, district
ORDER BY total_transactions DESC
LIMIT 5;
"""

df_top_districts_txn_latest = pd.read_sql(query_top_districts_txn_latest, conn)
df_top_districts_txn_latest

states                  district  total_transactions
0    karnataka  bengaluru urban district          3050041216
1    telangana        hyderabad district          2971669634
2  maharashtra             pune district          1994176959
3    rajasthan           jaipur district          1359854306
4        delhi          central district           811445152

In [103]:
import sqlite3
conn = sqlite3.connect('phonepe.db')

In [1]:
import sqlite3
import pandas as pd

conn = sqlite3.connect("phonepe.db")

In [2]:
pd.read_sql("PRAGMA table_info(state_yearly_transactions);", conn)

cid                name     type  notnull dflt_value  pk
0    0              states     TEXT        0       None   0
1    1                year  INTEGER        0       None   0
2    2  total_transactions  INTEGER        0       None   0
3    3        total_amount     REAL        0       None   0

In [3]:
pd.read_sql("PRAGMA table_info(state_trans_user);", conn)

cid          name     type  notnull dflt_value  pk
0    0        states     TEXT        0       None   0
1    1          year  INTEGER        0       None   0
2    2       quarter  INTEGER        0       None   0
3    3    trans_type     TEXT        0       None   0
4    4  trans_counts  INTEGER        0       None   0
5    5        amount     REAL        0       None   0
6    6         brand     TEXT        0       None   0
7    7   user_counts  INTEGER        0       None   0
8    8    percentage     REAL        0       None   0

In [4]:
pd.read_sql("PRAGMA table_info(district_transactions);", conn)

cid                    name     type  notnull dflt_value  pk
0    0                  states     TEXT        0       None   0
1    1                    year  INTEGER        0       None   0
2    2                 quarter  INTEGER        0       None   0
3    3                district     TEXT        0       None   0
4    4            trans_counts  INTEGER        0       None   0
5    5                  amount     REAL        0       None   0
6    6  registered_user_counts  INTEGER        0       None   0

In [5]:
pd.read_sql("PRAGMA table_info(top_transactions);", conn)

cid                    name     type  notnull dflt_value  pk
0    0                  states     TEXT        0       None   0
1    1                    year  INTEGER        0       None   0
2    2                 quarter  INTEGER        0       None   0
3    3                district     TEXT        0       None   0
4    4            trans_counts  INTEGER        0       None   0
5    5                  amount     REAL        0       None   0
6    6  registered_user_counts  INTEGER        0       None   0

In [6]:
import sqlite3
import pandas as pd

conn = sqlite3.connect("phonepe.db")
df = pd.read_sql("SELECT DISTINCT states FROM state_trans_user ORDER BY states;", conn)
df

states
0                andaman and nicobar islands
1                             andhra pradesh
2                          arunachal pradesh
3                                      assam
4                                      bihar
5                                 chandigarh
6                               chhattisgarh
7   dadra and nagar haveli and daman and diu
8                                      delhi
9                                        goa
10                                   gujarat
11                                   haryana
12                          himachal pradesh
13                         jammu and kashmir
14                                 jharkhand
15                                 karnataka
16                                    kerala
17                                    ladakh
18                               lakshadweep
19                            madhya pradesh
20                               maharashtra
21                                   manipur
22                                 meghalaya
23                                   mizoram
24                                  nagaland
25                                    odisha
26                                puducherry
27                                    punjab
28                                 rajasthan
29                                    sikkim
30                                tamil nadu
31                                 telangana
32                                   tripura
33                             uttar pradesh
34                               uttarakhand
35                               west bengal